packages/core/src/stdlib/modules/listModule.ts

```diff
@@ -303,6 +304,60 @@ export const listModule = InterpreterModule.create(
                     }
                 )
             ),
+            id(listProto).assignField(
+                "toString",
+                jsFun(
+                    (args, context) => {
+                        const self = args.getFieldValue(SemanticFieldNames.SELF, context);
+                        const length = assertNumber(self.getFieldValue(lengthField, context), "length field of a list");
+                        const maxDepthObject = args.getField("maxDepth", context).value;
+                        const maxDepth = maxDepthObject.isNull ? 4 : assertNumber(maxDepthObject);
+                        const fields: string[] = [];
+                        for (let i = 0; i < length; i++) {
+                            const next = self.getField(i, context).value;
+                            const nextCall = next.getFieldValue("toString", context);
+                            if (nextCall.isNull) {
+                                fields.push(next.toString(context, maxDepth - 1));
+                            } else {
+                                assertFunction(
+                                    nextCall,
+                                    `'toString()' is expected to be a function for ${next.toString(context, maxDepth - 1)}`
```

Comment:
this is eagerly evaluated
I would recommend to not have this here, due to its potential performance impact (in particular for large objects)

Reply:
I know, but I don't see a way around it unless we want the error message to be relatively hard to debug.

Reply:
extend the assert* functions to also take a callback?